forum/views/users.py

```diff
@@ -0,0 +1,433 @@
+"""Subscriptions API Views."""
+
+import math
+from typing import Any
+
+from rest_framework import status
+from rest_framework.exceptions import ParseError
+from rest_framework.permissions import AllowAny
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from forum.models import CommentThread, Contents, Users
+from forum.models.model_utils import (
+    find_or_create_user,
+    get_group_ids_from_params,
+    get_user_by_username,
+    handle_threads_query,
+    mark_as_read,
+    replace_username_in_all_content,
+    retire_all_content,
+    unsubscribe_all,
+    update_all_users_in_course,
+    user_to_hash,
+)
+from forum.serializers.thread import ThreadSerializer
+from forum.serializers.users import UserSerializer
+
+
+class UserAPIView(APIView):
+    """Users API View."""
+
+    permission_classes = (AllowAny,)
+
+    def get(self, request: Request, user_id: str) -> Response:
+        """Get user data."""
+        params: dict[str, Any] = request.GET.dict()
+        user = Users().get(user_id)
+        if not user:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+        group_ids = get_group_ids_from_params(params)
+        params.update({"group_ids": group_ids})
+        hashed_user = user_to_hash(user, params)
+        serializer = UserSerializer(hashed_user)
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    def put(self, request: Request, user_id: str) -> Response:
+        """Update user data."""
+        params = request.data
+        user = Users().get(user_id)
+        username = params.get("username")
+        user_by_username = get_user_by_username(username)
+        if user:
+            if (
+                user_by_username
+                and user["external_id"] != user_by_username["external_id"]
+            ):
+                return Response(status=status.HTTP_400_BAD_REQUEST)
+            Users().update(
+                user_id,
+                username=username,
+                default_sort_key=params.get("default_sort_key"),
+            )
+        else:
+            if user_by_username:
+                return Response(status=status.HTTP_400_BAD_REQUEST)
+            new_user_id = find_or_create_user(user_id)
+            Users().update(
+                new_user_id,
+                username=username,
+                default_sort_key=params.get("default_sort_key"),
+            )
```

Comment:
simplified it. Plus we can also add the validations for the params as username should be the required parameter, if we are going to create a new user.
```
if user and user_by_username:
    if user["external_id"] != user_by_username["external_id"]:
        return Response(status=status.HTTP_400_BAD_REQUEST)
elif user_by_username:
    return Response(status=status.HTTP_400_BAD_REQUEST)
else:
    user_id = find_or_create_user(user_id)
Users().update(
    user_id,
    username=username,
    default_sort_key=params.get("default_sort_key"),
)
```

Reply:
made code cleaner on above suggestion.

seems like username is not a required field for creating user as evident from this line in cs_comments_service,
```user = User.find_or_create_by(external_id: user_id)```  
Also, this api is used to update `default_sort_key` as well without requiring username.

Reply:
Actually i had to say while updating the user, we are passing data from params, without validating the username. Just cut the `if we are going to create a new user` part.